cli/js/tests/rename_test.ts

```diff
@@ -66,19 +80,137 @@ unitTest(
+  function renameSyncErrorsUnix(): void {
+    const testDir = Deno.makeTempDirSync();
+    const oldfile = testDir + "/oldfile";
+    const olddir = testDir + "/olddir";
+    const emptydir = testDir + "/empty";
+    const fulldir = testDir + "/dir";
+    const file = fulldir + "/file";
+    writeFileString(oldfile, "Hello");
+    Deno.mkdirSync(olddir);
+    Deno.mkdirSync(emptydir);
+    Deno.mkdirSync(fulldir);
+    writeFileString(file, "world");
+
+    assertThrows(
+      (): void => {
+        Deno.renameSync(oldfile, emptydir);
+      },
+      Error,
+      "Is a directory"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, fulldir);
+      },
+      Error,
+      "Directory not empty"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, file);
+      },
+      Error,
+      "Not a directory"
+    );
+
+    const fileLink = testDir + "/fileLink";
+    const dirLink = testDir + "/dirLink";
+    const danglingLink = testDir + "/danglingLink";
+    Deno.symlinkSync(file, fileLink);
+    Deno.symlinkSync(emptydir, dirLink);
+    Deno.symlinkSync(testDir + "/nonexistent", danglingLink);
+
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, fileLink);
+      },
+      Error,
+      "Not a directory"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, dirLink);
+      },
+      Error,
+      "Not a directory"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, danglingLink);
+      },
+      Error,
+      "Not a directory"
+    );
+
+    // should succeed on Unix
+    Deno.renameSync(olddir, emptydir);
+    Deno.renameSync(oldfile, dirLink);
+    Deno.renameSync(dirLink, danglingLink);
+    assertFile(danglingLink);
+    assertEquals("Hello", readFileString(danglingLink));
+  }
+);
+
+unitTest(
+  { ignore: Deno.build.os !== "win", perms: { read: true, write: true } },
+  function renameSyncErrorsWin(): void {
+    const testDir = Deno.makeTempDirSync();
+    const oldfile = testDir + "/oldfile";
+    const olddir = testDir + "/olddir";
+    const emptydir = testDir + "/empty";
+    const fulldir = testDir + "/dir";
+    const file = fulldir + "/file";
+    writeFileString(oldfile, "Hello");
+    Deno.mkdirSync(olddir);
+    Deno.mkdirSync(emptydir);
+    Deno.mkdirSync(fulldir);
+    writeFileString(file, "world");
+
+    assertThrows(
+      (): void => {
+        Deno.renameSync(oldfile, emptydir);
+      },
+      Deno.errors.PermissionDenied,
+      "Access is denied"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, fulldir);
+      },
+      Deno.errors.PermissionDenied,
+      "Access is denied"
+    );
+    assertThrows(
+      (): void => {
+        Deno.renameSync(olddir, emptydir);
+      },
+      Deno.errors.PermissionDenied,
+      "Access is denied"
+    );
+
+    // should succeed on Windows
+    Deno.renameSync(olddir, file);
+    assertDirectory(file);
```

Comment:
Nice tests

Reply:
Good, I'm glad you agree these are useful.